opnode/bss/config.go

```diff
@@ -0,0 +1,224 @@
+	L1EthRpc string
+
+	// L2EthRpc is the HTTP provider URL for L1.
+	L2EthRpc string
+
+	// CTCAddress is the CTC contract address.
+	CTCAddress string
+
+	// SROAddress is the SCC contract address.
+	SROAddress string
+
+	// MinL1TxSize is the minimum size in bytes of any L1 transactions generated
+	// by the batch submitter.
+	MinL1TxSize uint64
+
+	// MaxL1TxSize is the maximum size in bytes of any L1 transactions generated
+	// by the batch submitter.
+	MaxL1TxSize uint64
+
+	// MaxTxBatchCount is the maximum number of L2 transactions that can ever be
+	// in a batch.
+	MaxTxBatchCount uint64
+
+	// MaxStateBatchCount is the maximum number of L2 state roots that can ever
+	// be in a batch.
+	MaxStateBatchCount uint64
+
+	// MaxBatchSubmissionTime is the maximum amount of time that we will
+	// wait before submitting an under-sized batch.
+	MaxBatchSubmissionTime time.Duration
+
+	// PollInterval is the delay between querying L2 for more transaction
+	// and creating a new batch.
+	PollInterval time.Duration
+
+	// NumConfirmations is the number of confirmations which we will wait after
+	// appending new batches.
+	NumConfirmations uint64
+
+	// ResubmissionTimeout is time we will wait before resubmitting a
+	// transaction.
+	ResubmissionTimeout time.Duration
+
+	// FinalityConfirmations is the number of confirmations that we should wait
+	// before submitting state roots for CTC elements.
+	FinalityConfirmations uint64
+
+	// RunTxBatchSubmitter determines whether or not to run the tx batch
+	// submitter.
+	RunTxBatchSubmitter bool
+
+	// RunStateBatchSubmitter determines whether or not to run the state batch
+	// submitter.
+	RunStateBatchSubmitter bool
+
+	//SafeMinimumEtherBalance is the safe minimum amount of ether the batch
+	//submitter key should hold before it starts to log errors.
+	SafeMinimumEtherBalance uint64
+
+	// ClearPendingTxs is a boolean to clear the pending transactions in the
+	// mempool on startup.
+	ClearPendingTxs bool
+
+	/* Optional Params */
+
+	// LogLevel is the lowest log level that will be output.
+	LogLevel string
+
+	// BlockOffset is the offset between the CTC contract start and the L2 geth
+	// blocks.
+	BlockOffset uint64
+
+	// SequencerPrivateKey the private key of the wallet used to submit
+	// transactions to the CTC contract.
+	SequencerPrivateKey string
+
+	// PropopserPrivateKey the private key of the wallet used to submit
+	// transaction to the SCC contract.
+	L2OutputPrivateKey string
+
+	// Mnemonic is the HD seed used to derive the wallet private keys for both
+	// the sequence and l2output. Must be used in conjunction with
+	// SequencerHDPath and L2OutputHDPath.
+	Mnemonic string
+
+	// SequencerHDPath is the derivation path used to obtain the private key for
+	// the sequencer transactions.
+	SequencerHDPath string
+
+	// L2OutputHDPath is the derivation path used to obtain the private key for
+	// the l2output transactions.
+	L2OutputHDPath string
```

Comment:
This is how we configure the existing batch submitter. Using an encrypted keystore (https://github.com/roynalnaruto/eth-keystore-rs) could help to improve our security practices

Reply:
I wish that encrypted keystores worked with mnemonics by holding information about the hd path as part of its datastructure. Or if the encrypted keystore could hold a mnemonic and then the app could choose to derive whatever key it wants

Reply:
> Why do we import a mnemonic and specify derivation paths? I think we should instead specify a file-path towards an encrypted keystore and another to a keystore password, and mount/protect that password file with some key vault when deployed in production. IMHO mnemonics should only ever be used outside of deployed code, for backups etc. and never for key inputs.

I don't know the specifics of how it's implemented but in prod we use Google's secret manager to inject these. I have to imagine under the hood the setup is fairly similar to what you describe here—the mnemonic is not stored in a config file if that's your concern. As far using a raw key vs mnemonic+hdpath I don't have a strong preference. In the current BSS we use the mnemonic since it allows us to derive a key for each service (tx and state-roots) using only 1 pice of secret data, rather than multiple.